dali/util/fits.cc

```diff
@@ -115,11 +128,131 @@ void ParseHeader(HeaderData &parsed_header, fitsfile *src) {
+      mfpixel[i] = (long)fpixel[i];
+      mlpixel[i] = (long)lpixel[i];
+    } else {
+      mfpixel[i] = (long)lpixel[i];
+      mlpixel[i] = (long)fpixel[i];
+    }
+
+    naxis[i] = (fptr->Fptr)->znaxis[i];
+    tilesize[i] = (fptr->Fptr)->tilesize[i];
+    tiledim[i] = (naxis[i] - 1) / tilesize[i] + 1;
+    ftile[i] = (mfpixel[i] - 1) / tilesize[i] + 1;
+    ltile[i] = minvalue((mlpixel[i] - 1) / tilesize[i] + 1, tiledim[i]);
+    rowdim[i] = ntemp;
+    ntemp *= tiledim[i];
+  }
+
+  size = 0;
+  sum_nelemll = 0;
+  for (i5 = ftile[5]; i5 <= ltile[5]; ++i5) {
+    tfpixel[5] = (i5 - 1) * tilesize[5] + 1;
+    tlpixel[5] = minvalue(tfpixel[5] + tilesize[5] - 1, naxis[5]);
+    thistilesize[5] = tlpixel[5] - tfpixel[5] + 1;
+    offset[5] = (i5 - 1) * rowdim[5];
+    for (i4 = ftile[4]; i4 <= ltile[4]; ++i4) {
+      tfpixel[4] = (i4 - 1) * tilesize[4] + 1;
+      tlpixel[4] = minvalue(tfpixel[4] + tilesize[4] - 1, naxis[4]);
+      thistilesize[4] = thistilesize[5] * (tlpixel[4] - tfpixel[4] + 1);
+      offset[4] = (i4 - 1) * rowdim[4] + offset[5];
+      for (i3 = ftile[3]; i3 <= ltile[3]; ++i3) {
+        tfpixel[3] = (i3 - 1) * tilesize[3] + 1;
+        tlpixel[3] = minvalue(tfpixel[3] + tilesize[3] - 1, naxis[3]);
+        thistilesize[3] = thistilesize[4] * (tlpixel[3] - tfpixel[3] + 1);
+        offset[3] = (i3 - 1) * rowdim[3] + offset[4];
+        for (i2 = ftile[2]; i2 <= ltile[2]; ++i2) {
+          tfpixel[2] = (i2 - 1) * tilesize[2] + 1;
+          tlpixel[2] = minvalue(tfpixel[2] + tilesize[2] - 1, naxis[2]);
+          thistilesize[2] = thistilesize[3] * (tlpixel[2] - tfpixel[2] + 1);
+          offset[2] = (i2 - 1) * rowdim[2] + offset[3];
+          for (i1 = ftile[1]; i1 <= ltile[1]; ++i1) {
+            tfpixel[1] = (i1 - 1) * tilesize[1] + 1;
+            tlpixel[1] = minvalue(tfpixel[1] + tilesize[1] - 1, naxis[1]);
+            thistilesize[1] = thistilesize[2] * (tlpixel[1] - tfpixel[1] + 1);
+            offset[1] = (i1 - 1) * rowdim[1] + offset[2];
+            for (i0 = ftile[0]; i0 <= ltile[0]; ++i0) {
+              tfpixel[0] = (i0 - 1) * tilesize[0] + 1;
+              tlpixel[0] = minvalue(tfpixel[0] + tilesize[0] - 1, naxis[0]);
+              thistilesize[0] = thistilesize[1] * (tlpixel[0] - tfpixel[0] + 1);
+              irow = i0 + offset[1];
+
+              LONGLONG nelemll = 0, noffset = 0;
+              ffgdesll(fptr, (fptr->Fptr)->cn_compressed, irow, &nelemll, &noffset, status);
+
+              tile_offset[size] = sum_nelemll;
+              tile_size[size] = (long)thistilesize[0];
+
+              unsigned char charnull = 0;
+              raw_data[size].resize(nelemll / sizeof(unsigned char));
+              fits_read_col(fptr, TBYTE, (fptr->Fptr)->cn_compressed, irow, 1, (long)nelemll,
+                            &charnull, raw_data[size].data(), nullptr, status);
+
+              ++size;
+              sum_nelemll += nelemll;
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
Can this be converted to two nested loop and ask the compiler to unroll it (or just add a comment why it is unrolled by hand)?

Reply:
I added explanation comment. I do not think it can be converted to two nested loop and still support decompression up to six dimensions of FITS image. Does it make sense to use here #pragma GCC unroll N?

Reply:
@klecki 
Do you think that something like could work here or the compiler wouldn't inline it?
```
template <unsigned level>
inline void foo(tfpixel, tilesize, minvalue, rowdim, naxis, max_level=level) {
    for (i = ftile[level]; i <= ltile[level]; ++i) {
        tfpixel[level] = (i - 1) * tilesize[level] + 1;
        tlpixel[level] = minvalue(tfpixel[level] + tilesize[level] - 1, naxis[level]);
        if (level == max_level) {
            thistilesize[level] = tlpixel[level] - tfpixel[level] + 1;
            offset[level] = (i - 1) * rowdim[level];
        } else {
            thistilesize[level] = thistilesize[level + 1] * (tlpixel[level] - tfpixel[level] + 1);
            offset[level] = (i - 1) * rowdim[level] + offset[level+1];
        }
        foo<level-1>(tfpixel, tilesize, minvalue, rowdim, naxis, max_level);
    }    
}

template <>
inline void foo<0>(tfpixel, tilesize, minvalue, rowdim, naxis, max_level=0) {
    for (i = ftile[level]; i <= ltile[level]; ++i) {
        irow = i + offset[1];

        LONGLONG nelemll = 0, noffset = 0;
        ffgdesll(fptr, (fptr->Fptr)->cn_compressed, irow, &nelemll, &noffset, status);

        tile_offset[size] = sum_nelemll;
        tile_size[size] = (long)thistilesize[0];

        unsigned char charnull = 0;
        raw_data[size].resize(nelemll / sizeof(unsigned char));
        fits_read_col(fptr, TBYTE, (fptr->Fptr)->cn_compressed, irow, 1, (long)nelemll,
                    &charnull, raw_data[size].data(), nullptr, status);

        ++size;
        sum_nelemll += nelemll;
    }
}
foo<5>(tfpixel, tilesize, minvalue, rowdim, naxis);
```

Reply:
Sure, that should work.